### PR TITLE
[JENKINS-45294] - RemoteInvocationHandler#RPCRequest is now a subject for channel status checks

### PR DESCRIPTION
**Pull request by oleg-nenashev**
It is same as https://github.com/jenkinsci/remoting/pull/175, but for RPC Requests.
The idea is to prevent hanging of RPC Requests in edge cases when the channel goes down.

- [x] - Draft implementation
- [x] - Functional tests

https://issues.jenkins-ci.org/browse/JENKINS-45294

@reviewbybees @jglick

### CODE REVIEW

File: src/main/java/hudson/remoting/RemoteInvocationHandler.java [hunk 1]
```diff
@@ -118,26 +120,60 @@
...
+
+
+    /**
+     * Wraps an OID to the user-spaced typed wrapper.
+     *
+     * RPC operations with this object will be checking the channel state before sending the request.
+     * @see UserRPCRequest
+     */
+    @Nonnull
+    public static <T> T wrapUser(Channel channel, int id, Class<T> type, boolean userProxy, boolean autoUnexportByCaller) {
```
Comment (jglick) [inline src/main/java/hudson/remoting/RemoteInvocationHandler.java]: `public` modifier could be dropped, for clarity.

File: src/main/java/hudson/remoting/RemoteInvocationHandler.java [hunk 2]
```diff
@@ -118,26 +120,60 @@
...
+    }
+
+    /**
+     * Wraps an OID to the system-spaced typed wrapper.
+     *
+     * This method is restricted to internal {@link Channel}-specific use only.
+     * Usage of this method in external logic or on the {@link VirtualChannel} level is not supported.
+     */
+    @Nonnull
+    public static <T> T wrapSystem(Channel channel, int id, Class<T> type, boolean userProxy, boolean autoUnexportByCaller) {
```
Comment (jglick) [inline src/main/java/hudson/remoting/RemoteInvocationHandler.java]: ditto

File: src/main/java/hudson/remoting/RemoteInvocationHandler.java [hunk 3]
```diff
@@ -118,26 +120,60 @@
...
+        return internalWrap(channel, id, type, userProxy, autoUnexportByCaller, false);
     }
 
     /**
      * Wraps an OID to the typed wrapper.
      */
     @Nonnull
-    public static <T> T wrap(Channel channel, int id, Class<T> type, boolean userProxy, boolean autoUnexportByCaller) {
+    private static <T> T internalWrap(Channel channel, int id, Class<T> type, boolean userProxy, boolean autoUnexportByCaller, boolean userSpace) {
```
Comment (jglick) [inline src/main/java/hudson/remoting/RemoteInvocationHandler.java]: For use from `Channel.export` it would be simpler to just make this the only method, and delete the `wrapSystem` and `wrapUser` convenience methods.

File: src/main/java/hudson/remoting/RemoteInvocationHandler.java [hunk 4]
```diff
@@ -825,17 +861,17 @@ private void onChannelTermination(Channel channel) {
      * The downside of this is that the classes used as a parameter/return value
      * must be available to both JVMs.
      *
-     * If used as {@link Callable} in conjunction with {@link UserRequest},
-     * this can be used to send a method call to user-level objects, and
-     * classes for the parameters and the return value are sent remotely if needed.
+     * For user-space commands and operations, there is a {@link UserRPCRequest} implementation.
+     *
+     * @see UserRPCRequest
```
Comment (jglick) [inline src/main/java/hudson/remoting/RemoteInvocationHandler.java]: redundant, already `link`ed to

File: src/main/java/hudson/remoting/RemoteInvocationHandler.java [hunk 5]
```diff
@@ -935,12 +971,47 @@ private Method choose(Class[] interfaces) {
...
+     *
+     * This is an equivalent of {@link UserRequest} for RPC calls.
+     * Such kind of requests will not be send over closing or malfunctional channel.
+     *
+     * If used as {@link Callable} in conjunction with {@link UserRequest},
+     * this can be used to send a method call to user-level objects, and
+     * classes for the parameters and the return value are sent remotely if needed.
+     */
+    static class UserRPCRequest extends RPCRequest {
+        public UserRPCRequest(int oid, Method m, Object[] arguments, ClassLoader cl) {
```
Comment (jglick) [inline src/main/java/hudson/remoting/RemoteInvocationHandler.java]: I think you forgot to actually use this! Line 292.

File: src/main/java/hudson/remoting/RemoteInvocationHandler.java [hunk 6]
```diff
@@ -118,26 +120,60 @@
     private final Throwable origin;
 
     /**
+     * Indicates that the handler operates in the user space.
+     * In such case the requests will be automatically failed if the
+     * Remoting channel is not fully operational.
+     */
+    private final boolean userSpace;
```
Comment (jglick) [inline src/main/java/hudson/remoting/RemoteInvocationHandler.java]: Currently unused; see below.

File: src/main/java/hudson/remoting/RemoteInvocationHandler.java (hunk 5 above)
Comment (oleg-nenashev) [reply src/main/java/hudson/remoting/RemoteInvocationHandler.java]: oh my :( I will make sure tests actually catch it

File: src/main/java/hudson/remoting/RemoteInvocationHandler.java (hunk 6 above)
Comment (oleg-nenashev) [reply src/main/java/hudson/remoting/RemoteInvocationHandler.java]: fixed

File: src/main/java/hudson/remoting/RemoteInvocationHandler.java (hunk 5 above)
Comment (oleg-nenashev) [reply src/main/java/hudson/remoting/RemoteInvocationHandler.java]: done

**Comment by oleg-nenashev** [comment]
@jglick ready for re-review
